File: board.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>

#include "board.h"

static void clearBoard(int lines, int collumns, int board[][64]);
static int randomInteger(int low, int high);
static void insertBomb(int line, int collumn, int board[][64]);
static void increaseTile(int line, int collumn, int board[][64]);

/* Here I use a [lines+2][collumns+2] board with valid entries [1..lines][1..collumns].
 * This makes it easier to use functions such as insertBomb without corner verifications. */

void generateBoard(int lines, int collumns, int mines, int board[][64]) {
	int m;
	int lin, col;
	
	clearBoard(lines, collumns, board);
	
	srand(time(NULL));
	
	for (m = 0; m < mines; m++) {
		do {
			lin = randomInteger(1, lines);
			col = randomInteger(1, collumns);
		} while (board[lin][col] == -1);
		insertBomb(lin, col, board);
	}
}

static void clearBoard(int lines, int collumns, int board[][64]) {
	int i, j;
	for (i = 0; i <= lines + 1; i++)
		for (j = 0; j <= collumns + 1; j++)
			board[i][j] = 0;
}

/* randomInteger copied from Paulo Feofiloff, who copied himself from Eric Roberts */
static int randomInteger(int low, int high) {
    int k;
    double d;
    d = (double) rand( ) / ((double) RAND_MAX + 1);
    k = d * (high - low + 1);
    return low + k;
}

static void insertBomb(int line, int collumn, int board[][64]) {
	board[line][collumn] = -1;
	increaseTile(line-1, collumn-1, board);
	increaseTile(line-1, collumn,   board);
	increaseTile(line-1, collumn+1, board);
	increaseTile(line,   collumn-1, board);
	increaseTile(line,   collumn+1, board);
	increaseTile(line+1, collumn-1, board);
	increaseTile(line+1, collumn,   board);
	increaseTile(line+1, collumn+1, board);
}

static void increaseTile(int line, int collumn, int board[][64]) {
	if (board[line][collumn] == -1) return;
	board[line][collumn] += 1;
}

File: OSX/TTS Minesweeper Generator/Board.swift
//
//  Board.swift
//  TTS Minesweeper Generator
//
//  Created by Berna on 09/03/15.
//  Copyright (c) 2015 Berna & Epic Eric. All rights reserved.
//

import Foundation

public  class Board{

    class func newBoard(lines: Int, collumns: Int) -> Support.Matrix{
        var i:Int, j:Int, board = Support.Matrix(rows: lines + 2, columns: collumns + 2);
        for (i = 1; i < lines; i++){
            for (j = 1; j < collumns; j++){
                board[i, j] = 0
            }
        }
        return board
    }
    
    public class func generateBoard(lines: Int, collumns: Int, mines: Int) -> Support.Matrix{
        var m = 0, lin = 0, col = 0
        
        var board = newBoard(lines, collumns: collumns)
        
        for (m = 0; m < mines; m++){
            do {
                lin = Int(arc4random_uniform(UInt32(lines))) + 1
                col = Int(arc4random_uniform(UInt32(collumns))) + 1
            } while (board[lin, col] == -1)
            board = insertBomb(lin, collumn: col, board: board)
        }
        return board
    }
    
    class func insertBomb(line: Int, collumn: Int, board: Support.Matrix) -> Support.Matrix{
        var newBoard = board
        newBoard[line, collumn] = -1
        for (var i = line - 1 ; i <= line + 1; i++){
            for (var j = collumn - 1; j <= collumn + 1; j++){
                newBoard = increaseTile(i, collumn: j, board: newBoard)
            }
        }
        return newBoard
    }
    
    class func increaseTile (line: Int, collumn: Int, board: Support.Matrix) -> Support.Matrix {
        if (board[line, collumn] == -1){
            return board
        }
        var newBoard = board
        newBoard[line, collumn] = newBoard[line, collumn] + 1
        return newBoard
    }

}